### ISSUE TITLE
bug: Local SSL Certificate now unable to download

### ISSUE THREAD

**Issue by olliebatch** [localstack/localstack#11428]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

This only started over the last few days, i believe it's unrelated to code and just an existing permission change. 
Now unable to access the local https endpoints that get created in api gateway. And i believe this might also impact using the localstack website to access resource browsers. As SSL connections now fail, with invalid certificates.

This is just running a standard run command of the docker localstack

`	2024-08-29T08:21:32.353 DEBUG --- [  MainThread] localstack.utils.ssl       : Attempting to download local SSL certificate file
	2024-08-29T08:21:34.872  INFO --- [  MainThread] localstack.utils.http      : Unable to download Github artifact from https://cdn.jsdelivr.net/gh/localstack/localstack-artifacts@master/local-certs/server.key to /var/lib/localstack/cache/server.test.pem: Failed to download https://cdn.jsdelivr.net/gh/localstack/localstack-artifacts@master/local-certs/server.key, response code 403 Traceback (most recent call last):
	  File "/opt/code/localstack/localstack-core/localstack/utils/http.py", line 290, in do_download
	    download(download_url, target_file, timeout=timeout, request_headers=request_headers)
	  File "/opt/code/localstack/localstack-core/localstack/utils/http.py", line 202, in download
	    raise Exception("Failed to download %s, response code %s" % (url, r.status_code))
	Exception: Failed to download https://cdn.jsdelivr.net/gh/localstack/localstack-artifacts@master/local-certs/server.key, response code 403`

### Expected Behavior

I expect that the download of SSL when starting local stack succeeds and therefore allows access to https routes locally

### How are you starting LocalStack?

With a `docker run` command

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

    docker run localstack/localstack

This fails during initial run

### Environment

```markdown
- OS: MacOS
- LocalStack:
  LocalStack version: 3.6.0
  LocalStack Docker image sha: fb8addaddc00
  LocalStack build date:
  LocalStack build git hash:
```

**Comment by paymand** [comment]
@olliebatch try version 3.7.0. There's a fix.

**Comment by josalmi-sc** [comment]
I noticed the same. I wonder if it has something to do with the `localhost.localstack.cloud` certificate issuing being against the Let's Encrypt best practices: https://letsencrypt.org/docs/certificates-for-localhost/#for-native-apps-talking-to-web-apps

> You might be tempted to work around these limitations by setting up a domain name in the global DNS that happens to resolve to 127.0.0.1 (for instance, localhost.example.com), getting a certificate for that domain name, shipping that certificate and corresponding private key with your native app, and telling your web app to communicate with https://localhost.example.com:8000/ instead of http://127.0.0.1:8000/. **Don’t do this. It will put your users at risk, and your certificate may get revoked.**

**Comment by olliebatch** [comment]
I can confirm that the issue is fixed in 3.7.0 but appreciate that @josalmi-sc comment should be considered for localstack in the future

**Comment by Anze1508** [comment]
Hi,

We are pleased to inform you that we have implemented changes that should resolve the issue with the revoked certificate. These changes are incorporated in LocalStack version 3.7.0. To update your LocalStack container image to the latest version, or specifically to version 3.7.0, you can pull the latest image by using the following command:
`docker pull localstack/localstack-pro:latest`

Or, if you prefer to use a tagged version:
`docker pull localstack/localstack-pro:3.7.0`

Should you have any more questions or require further assistance, please do not hesitate to contact us.

**Comment by emoses** [comment]
This broke our CI pipeline until we updated to the latest localstack version, and I imaging many, many other folks are affected.  It looks like localstack pulls this cert and key from `localstack-artifacts@master`, so updates to the tip of that repo will affect *all prior versions* that relied on it.

It would have been great if older versions pointed to a specifc version tag of `localstack-artifacts`, but clearly that ship has sailed. Is there anything stopping you from adding `local-certs/server.key` back to master to un-break all the older localstack versions that people are currently happily using?

**Comment by yeradis** [comment]
Suddenly our CI pipeline started to fail, its using version 2.3.2 and its failing because of certificate issues

**Comment by ninjascribble** [comment]
This broke our CI pipeline as well. In our case upgrading to 3.7.0 was only part of the fix — we also needed to change the hostname in docker-compose.yml.
```yml
  localstack:
    image: localstack/localstack:3.7.0
    ports:
      - "${LOCALSTACK_PORT:-4566}:4566"
    environment:
      - HOSTNAME=localhost.localstack.cloud
      - LOCALSTACK_HOST=localhost.localstack.cloud
    # etc...
```

**Comment by komarkovich** [comment]
This is related to the ongoing issue. Please find more information in #11454

**Comment by Anze1508** [comment]
Hi everyone.

We are reaching out with good news about the [localhost.localstack.cloud](http://localhost.localstack.cloud/) certificate issue being resolved.

The incident has been resolved with the latest LocalStack patched version 3.7.2. No further service degradation has been observed for the past 48 hours.

**Important:**
**New LocalStack version:** All relevant fixes are applied in the new patched version 3.7.2. Make sure to [update your LocalStack image](https://docs.localstack.cloud/references/docker-images/) with the latest version by using latest or 3.7.2 tag.
**Older LocalStack versions**: Fixes are not applied to older versions of LocalStack, so customers using these may still encounter certificate revocation issues. If you experience this, please update to the latest version or [follow the steps outlined in our documentation](https://docs.localstack.cloud/getting-started/faq/#how-do-i-resolve-ssl-issues-due-to-revoked-local-certificate-for-localhostlocalstackcloud).
**Clear cached certificates**: If you continue to experience certificate revocation issues with the latest LocalStack version, please ensure that cached certificates are cleared from your host machine.

Thank you for your patience.

**Comment by ipsi-apant** [comment]
Hi @Anze1508 , this issue still present.

```shell
Attaching to localstack
localstack  |
localstack  | LocalStack version: 3.7.2
localstack  | LocalStack build date: 2024-09-06
localstack  | LocalStack build git hash: a607fed91
localstack  |
localstack  | 2024-09-22T23:54:30.265  INFO --- [  MainThread] localstack.utils.ssl       : Unable to download local test SSL certificate from https://api.localstack.cloud/v1/proxy/localstack.cert.key?version=3.7.2 to /var/lib/localstack/cache/server.test.pem (using self-signed cert as fallback): MyHTTPSConnectionPool(host='api.localstack.cloud', port=443): Max retries exceeded with url: /v1/proxy/localstack.cert.key?version=3.7.2 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa6a71fb610>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

### My docker compose

```yaml
services:
  localstack:
    container_name: "localstack"
    image: localstack/localstack:3.7.2
    ports:
      - "4566-4599:4566-4599"
      - "${PORT_WEB_UI-8080}:${PORT_WEB_UI-8080}"
    environment:
      - PERSISTENCE=1
      - SERVICES=dynamodb,ssm,sqs,kms,web,lambda,secretsmanager,s3,cognito,events
      - DEBUG=${DEBUG- }
      # - DOCKER_HOST=unix:///var/run/docker.sock
      - LOCALSTACK_HOST=localstack
      - HOSTNAME=localstack
      - READY_SCRIPT_PATH=${ready_script_path}
    volumes:
      - "${LOCALSTACK_VOLUME_DIR:-../../volume}:/var/lib/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"
      - "${READY_SCRIPT_PATH:-./docker/localstack-scripts/}:/etc/localstack/init/ready.d"
```

**Comment by Anze1508** [comment]
Hi @ipsi-apant. Did you try to delete your cached certificate file, as mentioned [here](https://docs.localstack.cloud/getting-started/faq/#how-do-i-resolve-ssl-issues-due-to-revoked-local-certificate-for-localhostlocalstackcloud)?

**Comment by ipsi-apant** [comment]
> Hi @ipsi-apant. Did you try to delete your cached certificate file, as mentioned [here](https://docs.localstack.cloud/getting-started/faq/#how-do-i-resolve-ssl-issues-due-to-revoked-local-certificate-for-localhostlocalstackcloud)?

Hi @Anze1508 , yes I have cleared the entire `volume` directory before running `docker compose`. `LOCALSTACK_VOLUME_DIR`  wipes clean before startup. Following is the shell script I am running.

```shell
#!/bin/bash

echo "(i) Removing containers..."
docker rm -f $(docker ps -a| grep localstack | cut -f1 -d' ' )
echo "(i) Containers removed."

echo "(i) Cleaning volume."
rm -rf ../../volume/*
echo "(i) Data load scripts will start executing."
localstack_ready_script_path="./docker/localstack-scripts/"

ready_script_path=$localstack_ready_script_path docker compose -f ./docker/docker-compose.yml --project-directory ./ build
ready_script_path=$localstack_ready_script_path docker compose -f ./docker/docker-compose.yml --project-directory ./ up
```

**Comment by Anze1508** [comment]
Hi @ipsi-apant. Could you please remove the `LOCALSTACK_HOST=localstack` variable from your docker compose and try again? Or, use the docker-compose we provide in our guide [here. ](https://docs.localstack.cloud/getting-started/installation/#starting-localstack-with-docker-compose)

Additionally, please download the certificate by clicking on this link: https://api.localstack.cloud/v1/proxy/localstack.cert.key?version=3.7.2

**Comment by ipsi-apant** [comment]
> Hi @ipsi-apant. Could you please remove the `LOCALSTACK_HOST=localstack` variable from your docker compose and try again? Or, use the docker-compose we provide in our guide [here. ](https://docs.localstack.cloud/getting-started/installation/#starting-localstack-with-docker-compose)
> 
> Additionally, please download the certificate by clicking on this link: https://api.localstack.cloud/v1/proxy/localstack.cert.key?version=3.7.2

Hi @Anze1508 , after removing `LOCALSTACK_HOST` env var, the certs download error disappeared from console. Also, certs and keys are present in the `volume/cache` directory. It appears resolved now.

Thank you for looking into this. Appreciate your time.

**Comment by Jolley71717** [comment]
I'm still seeing this issue with 3.7.2
```
2024-10-10 14:42:25 
2024-10-10 14:42:25 LocalStack version: 3.7.2
2024-10-10 14:42:25 LocalStack build date: 2024-09-06
2024-10-10 14:42:25 LocalStack build git hash: a607fed91
2024-10-10 14:42:25 
2024-10-10 14:42:27 2024-10-10T20:42:27.692  INFO --- [  MainThread] localstack.utils.ssl       : Unable to download local test SSL certificate from https://api.localstack.cloud/v1/proxy/localstack.cert.key?version=3.7.2 to /var/lib/localstack/cache/server.test.pem (using self-signed cert as fallback): MyHTTPSConnectionPool(host='api.localstack.cloud', port=443): Max retries exceeded with url: /v1/proxy/localstack.cert.key?version=3.7.2 (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')))
2024-10-10 14:42:27 Ready.
```

I'm starting the container using kotlin and TestContainers
```kotlin
private val localstackImage: DockerImageName = DockerImageName.parse("localstack/localstack:3.7.2")

    @Container
    private val localStack =
        LocalStackContainer(localstackImage)
            .withServices(LocalStackContainer.Service.S3, LocalStackContainer.Service.KMS)
```

I've seen suggestions in other threads to check if curl could be the problem by running `docker run --rm -it curlimages/curl https://api.localstack.cloud/v1/proxy/localstack.cert.key\?version\=3.7.2`
It looks like there still might be a problem
```
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the webpage mentioned above.
```

**Comment by thrau** [comment]
hey @Jolley71717, looks like your local machine does not trust the SSL cert of api.localstack.cloud, so this is not really related to the localstack version. you could investigate further with `docker run --rm -it curlimages/curl -v https://api.localstack.cloud/health`

**Comment by stephenmontgomery** [comment]
The instructions for the mitigation of old versions doesn't work. We're stuck on 3.6.0 because the SSE-C stuff introduced on 3.7.0 breaks our current S3 stuff (we already used presigned urls, SSE-C). Still need to investigate on that - AWS SDK throwing unhelpful BadRequest errors.

Anyho...

The instructions to fix for older versions ie prior to 3.7.2 didn't work for us. Clearing our localstack volume with the cache certs was done, setting `SKIP_SSL_CERT_DOWNLOAD=1` was done - same issue. We can't not use https endpoints because AWS SDK bulks at it because we're using SSE-C and SSE-C without https isn't allowed by AWS SDK. 

```
  localstack:
    image: localstack/localstack:3.6.0
    hostname: localhost.localstack.cloud
    environment:
      - AWS_DEFAULT_REGION=us-east-1
      - SERVICES=sqs,sns,s3,kms
      - DEBUG=1
      - DISABLE_CORS_CHECKS=1
#      - SKIP_SSL_CERT_DOWNLOAD=1
    ports:
      - "4566:4566"
      - "4510-4559:4510-4559"  # external services port range
    volumes:
      - 'localstack:/var/lib/localstack"'
      - '/var/run/docker.sock:/var/run/docker.sock'
      - './docker/aws/init-sns-sqs.sh:/etc/localstack/init/ready.d/init-sns-sqs.sh'
      - './docker/aws/kms.sh:/etc/localstack/init/ready.d/kms.sh'
      - './docker/aws/s3.sh:/etc/localstack/init/ready.d/s3.sh'
```

I didn't understand why a self-cert would work - we needed a cert with `*.localhost.localstack.cloud` in there.

<img width="1115" alt="Image" src="https://github.com/user-attachments/assets/242238c5-dd91-46ef-8abd-b0494cd7d55b" />

Eventually I got it working with blowing away my localstack volume, temporarily used localstack/localstack:stable image to download the correct certs and then changed back to 3.6.0 to reuse the correct certs. Back working.

What I did notice was the flakiness of cert downloads on both 3.7.2 and stable - there was frequent "retry failed" attempting to download the certs and then quietly giving up (rate limited endpoint???) and just generating a self-cert as fallback which was no good. Eventually the stable version downloaded worked and I knew that it actually worked because my docker volume went from a few KBs (just the certs) to > 65mb because it downloaded JDK 11 for internal stuff:

<img width="1631" alt="Image" src="https://github.com/user-attachments/assets/24a73921-535a-481d-a22f-40cf866782a8" />

So, just writing this up to save other impacted people doing a deep dive into this.

**Comment by bentsku** [comment]
Thanks a lot @stephenmontgomery for the instructions!

> We're stuck on 3.6.0 because the SSE-C stuff introduced on 3.7.0 breaks our current S3 stuff (we already used presigned urls, SSE-C). Still need to investigate on that - AWS SDK throwing unhelpful BadRequest errors.

Regarding this issue, would you mind opening a new issue with one of the failing requests and logging with `LS_LOG=trace`? Are we sure it works against AWS? I'd be interested in looking into the issue and hopefully unblock you. Thank you!
